src/main/java/htsjdk/samtools/util/AsyncWriterPool.java

```diff
@@ -0,0 +1,245 @@
+         * @param queue      a queue to use for this writer, bounded or unbounded.
+         * @param bufferSize the minimum number of items to write needed before sending to a thread for writing.
+         * @throws IllegalArgumentException if bufferSize <= 0
+         */
+        public PooledWriter(final Writer<A> writer, final BlockingQueue<A> queue, final int bufferSize) {
+            if (bufferSize <= 0) throw new IllegalArgumentException("bufferSize must be >= 1");
+
+            this.writer = writer;
+            this.queue = queue;
+            this.bufferSize = bufferSize;
+            this.writing = new AtomicBoolean(false);
+            this.isClosed = false;
+            this.currentTask = null;
+
+            // Add to pools writers
+            AsyncWriterPool.this.writers.add(this);
+        }
+
+        /**
+         * Exchange a class implementing {@link Writer} for a {@code PooledWriter}.
+         * <p>
+         * This uses a {@link LinkedBlockingQueue} and a {@code bufferSize} of 10,000.
+         *
+         * @param writer a class implementing {@link Writer}
+         */
+        public PooledWriter(final Writer<A> writer) {
+            this(writer, new LinkedBlockingQueue<>(), 10_000);
+        }
+
+        /**
+         * Exchange a class implementing {@link Writer} for a {@code PooledWriter}.
+         * <p>
+         * This uses a {@link LinkedBlockingQueue}.
+         *
+         * @param writer     a class implementing {@link Writer}
+         * @param bufferSize the minimum number of items to write needed before sending to a thread for writing.
+         */
+        public PooledWriter(final Writer<A> writer, final int bufferSize) {
+            this(writer, new LinkedBlockingQueue<>(), bufferSize);
+        }
+
+        /**
+         * Update the minimum number of items required before sending items to a thread for writing.
+         *
+         * @param newBufferSize the new size to use for {@code bufferSize}
+         * @throws IllegalArgumentException if {@code newBufferSize} <= 0
+         */
+        public void setBufferSize(final int newBufferSize) {
+            if (newBufferSize <= 0) throw new IllegalArgumentException("bufferSize must be >= 1");
+            this.bufferSize = newBufferSize;
+        }
+
+
+        /**
+         * Will throw any error thrown in task
+         *
+         * @throws RuntimeException if any exception was raised during the writing on the receiving thread
+         */
+        private void checkAndRethrow() {
+            if (this.currentTask != null) {
+                try {
+                    // NB: ignoring output of number of records previously written
+                    this.currentTask.get();
+                } catch (CancellationException | InterruptedException | ExecutionException e) {
+                    this.isClosed = true;
+                    throw new RuntimeException("Exception while writing records asynchronously", e);
+                }
+            }
```

Comment:
I think there should probably be a `finally { this.currentTask = null }` in here so that we're not continuously checking the same task - we only need to check it once per `Future`.